File: frontend/src/lib/activity.ts
import type { PipelineEvent, Severity, StageId } from "../types";

/**
 * Modelo de la actividad del análisis a partir de los eventos del backend (función pura).
 * Para reproducir una sesión basta con construirlo con los eventos cuyo `t` ≤ cabezal.
 */

export const STAGE_ORDER: StageId[] = ["preparing", "auditing", "validating", "migration", "done"];

export interface Check {
  name: string;
  value: string;
  limit: string;
}

export interface Inventory {
  files: number;
  bytes: number;
  python_lines: number;
  languages: { name: string; files: number }[];
  top_dirs: { name: string; files: number }[];
  excluded: string[];
  sha256: string;
}

export interface AgentRun {
  name: string;
  task: string | null;
  status: "running" | "done";
  startedAt: number;
  endedAt: number | null;
  toolUses: number | null;
  turns: number | null;
  cost: number | null;
  durationMs: number | null;
  /** Primera frase del informe que el subagente devolvió al orquestador. */
  report: string | null;
}

export interface FeedItem {
  seq: number;
  t: number;
  kind: string;
  actor: string;
  title: string;
  detail: string | null;
  target: string | null;
}

export interface PlanItem {
  state: "done" | "active" | "pending";
  text: string;
}

export interface EvidenceCheckItem {
  seq: number;
  t: number;
  findingId: string;
  status: "valid" | "invalid";
  path: string | null;
  lineStart: number | null;
  lineEnd: number | null;
  severity: Severity | null;
  reason: string | null;
}

export interface TestRun {
  seq: number;
  target: "legacy" | "modern";
  name: string;
  status: "passed" | "failed" | "not_run";
  durationMs: number;
  reason: string | null;
}

export interface ActivityModel {
  stages: Record<StageId, { start: number | null; end: number | null; events: number }>;
  checks: Check[];
  inventory: Inventory | null;
  bob: {
    maxCost: number | null;
    exploreCost: number | null;
    subagentsEnabled: boolean | null;
    turns: number;
    tools: number;
    plan: PlanItem[];
    agents: AgentRun[];
    skills: string[];
    feed: FeedItem[];
    finalize: FeedItem | null;
    result: { cost: number | null; durationMs: number | null; toolCalls: number | null } | null;
    answered: boolean;
    recorded: boolean;
  };
  evidence: { checks: EvidenceCheckItem[]; summary: { accepted: number; rejected: number; valid: number; total: number; riskScored: number; pertDays: number | null } | null };
  migration: { tests: TestRun[]; skipped: string | null; summary: { status: string; endpoint: string } | null };
  done: { findings: number; bySeverity: Record<string, number>; evidenceValid: number; evidenceTotal: number; cost: number | null; durationMs: number | null } | null;
  failure: { stage: StageId; message: string } | null;
  lastT: number;
}

const num = (value: unknown): number | null => (typeof value === "number" && Number.isFinite(value) ? value : null);
const str = (value: unknown): string | null => (typeof value === "string" && value.length > 0 ? value : null);

function emptyModel(): ActivityModel {
  return {
    stages: Object.fromEntries(STAGE_ORDER.map((stage) => [stage, { start: null, end: null, events: 0 }])) as ActivityModel["stages"],
    checks: [],
    inventory: null,
    bob: { maxCost: null, exploreCost: null, subagentsEnabled: null, turns: 0, tools: 0, plan: [], agents: [], skills: [], feed: [], finalize: null, result: null, answered: false, recorded: false },
    evidence: { checks: [], summary: null },
    migration: { tests: [], skipped: null, summary: null },
    done: null,
    failure: null,
    lastT: 0,
  };
}

const FEED_KINDS = new Set(["bob.tool", "bob.thinking", "bob.skill", "bob.subagent.start", "bob.subagent.end", "bob.subagent.report", "bob.plan", "bob.finalize", "bob.answer", "bob.writing", "bob.tool.error"]);

export function buildActivity(events: PipelineEvent[]): ActivityModel {
  const model = emptyModel();
  for (const event of events) {
    const stage = model.stages[event.stage];
    if (stage) {
      stage.start = stage.start === null ? event.t : Math.min(stage.start, event.t);
      stage.end = Math.max(stage.end ?? event.t, event.t);
      stage.events += 1;
    }
    model.lastT = Math.max(model.lastT, event.t);
    if (event.recorded) model.bob.recorded = true;
    apply(model, event);
  }
  return model;
}

/** Reemplaza (sin mutarlo) el primer subagente que cumple la condición; false si no hay ninguno. */
function patchAgent(model: ActivityModel, match: (agent: AgentRun) => boolean, patch: Partial<AgentRun>): boolean {
  const index = model.bob.agents.findIndex(match);
  if (index < 0) return false;
  model.bob.agents = model.bob.agents.map((agent, i) => (i === index ? { ...agent, ...patch } : agent));
  return true;
}

function apply(model: ActivityModel, event: PipelineEvent): void {
  const d = event.data;
  if (FEED_KINDS.has(event.kind)) {
    const chars = num(d.chars);
    const title = (event.kind === "bob.writing" || event.kind === "bob.answer" || event.kind === "bob.subagent.report") && chars
      ? `${event.title} · ${chars.toLocaleString("es")} caracteres`
      : event.title;
    model.bob.feed.push({ seq: event.seq, t: event.t, kind: event.kind, actor: event.actor, title, detail: event.detail, target: str(d.target) });
  }
  switch (event.kind) {
    case "ingest.check":
      model.checks.push({ name: event.title, value: str(d.value) ?? "", limit: str(d.limit) ?? "" });
      break;
    case "inventory":
      model.inventory = {
        files: num(d.files) ?? 0,
        bytes: num(d.bytes) ?? 0,
        python_lines: num(d.python_lines) ?? 0,
        languages: Array.isArray(d.languages) ? (d.languages as Inventory["languages"]) : [],
        top_dirs: Array.isArray(d.top_dirs) ? (d.top_dirs as Inventory["top_dirs"]) : [],
        excluded: Array.isArray(d.excluded) ? (d.excluded as string[]) : [],
        sha256: str(d.sha256) ?? "",
      };
      break;
    case "bob.start":
      model.bob.maxCost = num(d.max_cost);
      model.bob.exploreCost = num(d.explore_cost);
      model.bob.subagentsEnabled = typeof d.subagents === "boolean" ? d.subagents : null;
      break;
    case "bob.turn":
      model.bob.turns = Math.max(model.bob.turns, num(d.turn) ?? 0);
      break;
    case "bob.tool":
      model.bob.tools += 1;
      break;
    case "bob.skill":
      if (str(d.skill) && !model.bob.skills.includes(d.skill as string)) model.bob.skills.push(d.skill as string);
      break;
    case "bob.plan":
      if (Array.isArray(d.items)) model.bob.plan = d.items as PlanItem[];
      break;
    case "bob.subagent.start":
      model.bob.agents.push({
        name: str(d.agent) ?? "subagente", task: event.detail, status: "running", startedAt: event.t,
        endedAt: null, toolUses: null, turns: null, cost: null, durationMs: null, report: null,
      });
      break;
    // Bob no da un id al iniciar un subagente: cada informe y cada cierre llenan el primer subagente
    // con ese nombre que aún no lo tiene, sin sobrescribir nunca uno ya completo.
    case "bob.subagent.report": {
      const name = str(d.agent) ?? "subagente";
      const report = event.detail ?? "";
      if (!patchAgent(model, (agent) => agent.name === name && agent.report === null, { report })) {
        // Sin el log de Bob no hubo subagent_start: el informe basta para saber que trabajó.
        model.bob.agents = [...model.bob.agents, {
          name, task: null, status: "done", startedAt: event.t, endedAt: event.t,
          toolUses: null, turns: null, cost: null, durationMs: null, report,
        }];
      }
      break;
    }
    case "bob.subagent.end": {
      const name = str(d.agent) ?? "subagente";
      patchAgent(model, (agent) => agent.name === name && agent.status === "running", {
        status: "done", endedAt: event.t, toolUses: num(d.tool_uses), turns: num(d.turns), cost: num(d.cost), durationMs: num(d.duration_ms),
      });
      break;
    }
    case "bob.finalize":
      model.bob.finalize = model.bob.feed[model.bob.feed.length - 1] ?? null;
      break;
    case "bob.answer":
      model.bob.answered = true;
      break;
    case "bob.result":
      model.bob.result = { cost: num(d.cost), durationMs: num(d.duration_ms), toolCalls: num(d.tool_calls) };
      break;
    case "evidence.check":
      model.evidence.checks.push({
        seq: event.seq, t: event.t, findingId: str(d.finding_id) ?? "?", status: d.status === "valid" ? "valid" : "invalid",
        path: str(d.path), lineStart: num(d.line_start), lineEnd: num(d.line_end), severity: (str(d.severity) as Severity | null), reason: event.detail,
      });
      break;
    case "evidence.summary":
      model.evidence.summary = {
        accepted: num(d.accepted) ?? 0, rejected: num(d.rejected) ?? 0, valid: num(d.valid) ?? 0, total: num(d.total) ?? 0,
        riskScored: num(d.risk_scored) ?? 0, pertDays: num(d.pert_expected_days),
      };
      break;
    case "tests.result":
      model.migration.tests.push({
        seq: event.seq, target: d.target === "modern" ? "modern" : "legacy", name: str(d.name) ?? event.title,
        status: d.status === "passed" ? "passed" : d.status === "failed" ? "failed" : "not_run", durationMs: num(d.duration_ms) ?? 0, reason: event.detail,
      });
      break;
    case "tests.skipped":
      model.migration.skipped = event.detail ?? event.title;
      break;
    case "tests.summary":
      model.migration.summary = { status: str(d.status) ?? "", endpoint: str(d.endpoint) ?? "" };
      break;
    case "dossier.ready":
      model.done = {
        findings: num(d.findings) ?? 0, bySeverity: (d.by_severity as Record<string, number>) ?? {},
        evidenceValid: num(d.evidence_valid) ?? 0, evidenceTotal: num(d.evidence_total) ?? 0,
        cost: num(d.bob_cost), durationMs: num(d.bob_duration_ms),
      };
      break;
    case "pipeline.failed":
      model.failure = { stage: event.stage, message: event.detail ?? event.title };
      break;
  }
}

/** Resumen de una línea por etapa para el riel (solo cifras presentes en los eventos). */
export function stageMetric(model: ActivityModel, stage: StageId): string | null {
  switch (stage) {
    case "preparing":
      return model.inventory ? `${model.inventory.files} archivos · ${model.inventory.python_lines.toLocaleString("es")} líneas Python` : null;
    case "auditing": {
      const parts: string[] = [];
      if (model.bob.tools) parts.push(`${model.bob.tools} lecturas y búsquedas`);
      if (model.bob.agents.length) parts.push(`${model.bob.agents.length} subagente${model.bob.agents.length === 1 ? "" : "s"}`);
      if (model.bob.result?.cost != null) parts.push(`${model.bob.result.cost.toFixed(2)} bc`);
      return parts.length ? parts.join(" · ") : null;
    }
    case "validating": {
      if (model.evidence.summary) return `${model.evidence.summary.valid}/${model.evidence.summary.total} citas verificadas`;
      const valid = model.evidence.checks.filter((c) => c.status === "valid").length;
      return model.evidence.checks.length ? `${valid}/${model.evidence.checks.length} citas…` : null;
    }
    case "migration":
      if (model.migration.skipped) return "No se ejecuta código subido";
      return model.migration.tests.length
        ? `${model.migration.tests.filter((t) => t.status === "passed").length}/${model.migration.tests.length} pruebas pasan`
        : null;
    case "done":
      return model.done ? `${model.done.findings} hallazgos con evidencia` : null;
  }
}

export const formatClock = (seconds: number) => {
  const s = Math.max(0, Math.floor(seconds));
  return `${Math.floor(s / 60)}:${String(s % 60).padStart(2, "0")}`;
};

File: backend/app/pipeline/activity.py
"""Registro de actividad de un análisis: qué hace cada etapa mientras ocurre.

- `EventLog` escribe eventos (JSONL) en `artifacts/jobs/<id>/events.jsonl`; la API los sirve con cursor.
- `BobActivity` traduce el stream de `bob run --format stream-json` en eventos legibles: plan (todo list),
  herramientas, delegación en subagentes, razonamiento, turnos y coste.
- `inventory_events` describe el sandbox (lenguajes, líneas, carpetas) sin ejecutar nada.

Los eventos son un resumen saneado: nunca incluyen el contenido de los archivos que Bob lee, y las rutas
absolutas del servidor se convierten en rutas relativas al repositorio.
"""

import json
import re
import threading
import time
from collections import Counter
from datetime import datetime
from pathlib import Path
from typing import Any, Literal

from pydantic import BaseModel, Field

EVENTS_FILE = "events.jsonl"
Stage = Literal["preparing", "auditing", "validating", "migration", "done"]

MAX_TITLE = 200
MAX_DETAIL = 600
MAX_THINKING = 480
MAX_EVENTS_PER_PAGE = 400
MAX_EVENTS = 5000  # tope por análisis: por encima solo se registran eventos esenciales
MAX_TODOS = 40
ESSENTIAL_KINDS = frozenset({"stage.start", "pipeline.failed", "bob.result", "bob.finalize", "evidence.summary",
                             "tests.summary", "tests.skipped", "dossier.ready"})
_SEQ_PREFIX = re.compile(r'^\{"seq":(\d+),')
# Rutas absolutas POSIX o Windows (no URLs): se reducen a su último componente.
_ABSOLUTE_PATH = re.compile(r"(?<![\w.:/])/(?:[^\s/\"'`<>|:]+/)+([^\s/\"'`<>|:]*)|\b[A-Za-z]:\\(?:[^\s\\\"'`<>|]+\\)*([^\s\\\"'`<>|]*)")
WRITING_STEP = 2000  # cada cuántos caracteres del JSON final se informa el progreso
_TODO = re.compile(r"^\s*\[( |x|X|-)\]\s*(.+?)\s*$")
_TASK_TAGS = re.compile(r"</?task_result>")
MAX_REPORT = 220
_WHITESPACE = re.compile(r"\s+")
# Extensiones que se cuentan como código en el inventario (el resto se agrupa como "otros").
LANGUAGES = {
    ".py": "Python", ".sql": "SQL", ".html": "HTML", ".js": "JavaScript", ".jsx": "JavaScript",
    ".ts": "TypeScript", ".tsx": "TypeScript", ".css": "CSS", ".md": "Markdown", ".json": "JSON",
    ".yml": "YAML", ".yaml": "YAML", ".toml": "TOML", ".ini": "INI", ".txt": "Texto",
}


class PipelineEvent(BaseModel):
    seq: int
    t: float = Field(description="Segundos desde el inicio del análisis.")
    stage: Stage
    kind: str
    actor: str
    title: str
    detail: str | None = None
    data: dict[str, Any] = Field(default_factory=dict)
    recorded: bool = False


def redact_paths(text: str) -> str:
    """Nunca expone rutas del servidor: cualquier ruta absoluta queda solo con su nombre final."""
    return _ABSOLUTE_PATH.sub(lambda match: match.group(1) or match.group(2) or "…", text)


def _clip(text: str | None, limit: int) -> str | None:
    if text is None:
        return None
    text = _WHITESPACE.sub(" ", text).strip()
    return text if len(text) <= limit else text[: limit - 1] + "…"


class EventLog:
    """Registro append-only y seguro entre hilos. Un archivo JSONL por análisis."""

    def __init__(self, path: Path) -> None:
        self.path = path
        self._lock = threading.Lock()
        self._started = time.monotonic()
        self._offset = 0.0
        self._seq = self._last_seq()

    def _last_seq(self) -> int:
        if not self.path.is_file():
            return 0
        last = 0
        for line in self.path.read_text(encoding="utf-8").splitlines():
            try:
                last = max(last, int(json.loads(line)["seq"]))
            except (ValueError, KeyError, json.JSONDecodeError):
                continue
        return last

    def now(self) -> float:
        return round(time.monotonic() - self._started + self._offset, 3)

    def advance(self, seconds: float) -> None:
        """Desplaza el reloj (al insertar una sesión grabada, lo posterior va después de ella)."""
        with self._lock:
            self._offset += max(0.0, seconds)

    def emit(self, stage: Stage, kind: str, actor: str, title: str, detail: str | None = None,
             data: dict[str, Any] | None = None, t: float | None = None, recorded: bool = False) -> PipelineEvent | None:
        with self._lock:
            if self._seq >= MAX_EVENTS and kind not in ESSENTIAL_KINDS:
                return None  # sesión anómala: el archivo no crece sin límite
            self._seq += 1
            event = PipelineEvent(
                seq=self._seq, t=self.now() if t is None else round(t, 3), stage=stage, kind=kind, actor=actor,
                title=_clip(title, MAX_TITLE) or "", detail=_clip(detail, MAX_DETAIL), data=data or {}, recorded=recorded,
            )
            self.path.parent.mkdir(parents=True, exist_ok=True)
            with self.path.open("a", encoding="utf-8") as handle:
                handle.write(event.model_dump_json() + "\n")
        return event


def read_events(path: Path, after: int = 0, limit: int = MAX_EVENTS_PER_PAGE) -> list[PipelineEvent]:
    """Eventos con `seq > after`, en orden. Ignora líneas corruptas (escritura interrumpida)."""
    if not path.is_file():
        return []
    events: list[PipelineEvent] = []
    for line in path.read_text(encoding="utf-8").splitlines():
        prefix = _SEQ_PREFIX.match(line)
        if prefix and int(prefix.group(1)) <= after:
            continue  # ya entregado: no se valida de nuevo en cada sondeo
        try:
            event = PipelineEvent.model_validate_json(line)
        except ValueError:
            continue
        if event.seq > after:
            events.append(event)
        if len(events) >= limit:
            break
    return events


# --- Inventario del sandbox ---------------------------------------------------------------

def inventory_events(workspace: Path) -> dict[str, Any]:
    """Resumen medible del repositorio copiado al sandbox (sin la configuración .bob/)."""
    languages: Counter[str] = Counter()
    python_lines = 0
    files = 0
    total_bytes = 0
    top_dirs: Counter[str] = Counter()
    for path in workspace.rglob("*"):
        if not path.is_file():
            continue
        relative = path.relative_to(workspace)
        if relative.parts[0] == ".bob":
            continue
        files += 1
        total_bytes += path.stat().st_size
        languages[LANGUAGES.get(path.suffix.lower(), "Otros")] += 1
        top_dirs[relative.parts[0] if len(relative.parts) > 1 else "(raíz)"] += 1
        if path.suffix.lower() == ".py":
            try:
                python_lines += sum(1 for _ in path.open(encoding="utf-8", errors="replace"))
            except OSError:
                continue
    return {
        "files": files,
        "bytes": total_bytes,
        "python_lines": python_lines,
        "languages": [{"name": name, "files": count} for name, count in languages.most_common(8)],
        "top_dirs": [{"name": name, "files": count} for name, count in top_dirs.most_common(8)],
    }


# --- Stream de Bob ------------------------------------------------------------------------

def relative_to_workspace(value: str, workspace: Path) -> str:
    """Nunca expone rutas del servidor: relativas al repo, o solo el nombre si están fuera."""
    text = value.strip()
    root = str(workspace.resolve())
    if text.startswith(root):
        text = text[len(root):].lstrip("/\\") or "."
    elif text.startswith("/") or re.match(r"^[A-Za-z]:\\", text):
        text = Path(text).name
    return text


def _strip_paths(text: str, workspace: Path) -> str:
    root = str(workspace.resolve())
    return redact_paths(text.replace(root + "/", "").replace(root, "."))


def parse_todos(raw: str) -> list[dict[str, str]]:
    items: list[dict[str, str]] = []
    for line in raw.splitlines()[:MAX_TODOS * 4]:
        if len(items) >= MAX_TODOS:
            break
        match = _TODO.match(line)
        if match:
            mark = match.group(1).lower()
            state = "done" if mark == "x" else "active" if mark == "-" else "pending"
            items.append({"state": state, "text": _clip(match.group(2), 160) or ""})
    return items


class BobActivity:
    """Convierte eventos crudos de Bob (stream-json) en eventos de la etapa `auditing`.

    El intérprete es puro respecto al contenido: no guarda salidas de herramientas ni código leído.
    """

    ORCHESTRATOR = "evidence-auditor"

    def __init__(self, log: EventLog, workspace: Path, actor: str = ORCHESTRATOR,
                 recorded: bool = False, clock: float | None = None) -> None:
        self.log = log
        self.workspace = workspace
        self.actor = actor
        self.recorded = recorded
        self.clock = clock
        self._text: list[str] = []
        self._chars = 0
        self._reported = 0
        self._turn = 0
        self._last_tokens: tuple[int, int] | None = None
        self._first_ts: float | None = None
        self._spawned: dict[str, str] = {}
        # El stream (stdout) y el log de Bob (tiempo real) llegan desde hilos distintos.
        self._lock = threading.Lock()

    def _t(self, raw: dict[str, Any]) -> float | None:
        """En una sesión grabada se conserva el ritmo original a partir de `clock`."""
        if self.clock is None:
            return None
        try:
            seconds = datetime.fromisoformat(str(raw.get("timestamp", "")).replace("Z", "+00:00")).timestamp()
        except ValueError:
            return None
        if self._first_ts is None:
            self._first_ts = seconds
        return self.clock + (seconds - self._first_ts)

    def _emit(self, raw: dict[str, Any], kind: str, title: str, detail: str | None = None, data: dict[str, Any] | None = None,
              actor: str | None = None) -> None:
        self.log.emit("auditing", kind, actor or self.actor, title, detail, data, t=self._t(raw), recorded=self.recorded)

    def flush_thinking(self, raw: dict[str, Any] | None = None) -> None:
        text = "".join(self._text).strip()
        self._text = []
        self._chars = 0
        self._reported = 0
        if not text:
            return
        if text.lstrip().startswith("{") or '"findings"' in text[:200]:
            self._emit(raw or {}, "bob.answer", "Bob entregó el expediente", data={"chars": len(text)})
            return
        self._emit(raw or {}, "bob.thinking", "Razonamiento", _clip(_strip_paths(text, self.workspace), MAX_THINKING))

    def feed(self, raw: dict[str, Any]) -> None:
        with self._lock:
            self._feed(raw)

    def _feed(self, raw: dict[str, Any]) -> None:
        kind = raw.get("type")
        if kind == "message" and raw.get("role") == "assistant":
            chunk = str(raw.get("content", ""))
            self._text.append(chunk)
            self._chars += len(chunk)
            # Redactar el expediente puede tardar minutos: se informa el avance sin exponer el texto.
            if self._chars - self._reported >= WRITING_STEP and "".join(self._text).lstrip().startswith("{"):
                self._reported = self._chars - self._chars % WRITING_STEP
                self._emit(raw, "bob.writing", "Bob redacta el expediente", data={"chars": self._reported})
        elif kind == "tool_use":
            self.flush_thinking(raw)
            self._tool(raw)
        elif kind == "tool_result" and str(raw.get("tool_id")) in self._spawned:
            self._report(raw)
        elif kind == "tool_result" and raw.get("status") not in (None, "success"):
            self._emit(raw, "bob.tool.error", "Una herramienta falló", _clip(_strip_paths(str(raw.get("output", "")), self.workspace), 160))
        elif kind == "subagent_start":
            agent = str(raw.get("agentType", "subagente"))
            task = _clip(_strip_paths(str(raw.get("description", "")), self.workspace), 280)
            self._emit(raw, "bob.subagent.start", f"Delegó en {agent}", task, {"agent": agent})
        elif kind == "subagent_end":
            meta = raw.get("metadata") or {}
            agent = str(meta.get("agentType", "subagente"))
            spend = meta.get("spend") or {}
            self._emit(raw, "bob.subagent.end", f"{agent} terminó", data={
                "agent": agent,
                "tool_uses": meta.get("toolUseCount"),
                "turns": meta.get("loopTurnCount"),
                "duration_ms": meta.get("durationMs"),
                "cost": spend.get("cost"),
                "exit": meta.get("loopExitReason"),
            })
        elif kind == "cost":
            costs = raw.get("costs") or {}
            tokens = (int(costs.get("input", 0) or 0), int(costs.get("output", 0) or 0))
            if tokens != self._last_tokens:
                self._last_tokens = tokens
                self._turn += 1
                self._emit(raw, "bob.turn", f"Turno {self._turn} del orquestador",
                           data={"turn": self._turn, "input_tokens": tokens[0], "output_tokens": tokens[1]})
        elif kind == "result":
            self.flush_thinking(raw)
            stats = raw.get("stats") or {}
            self._emit(raw, "bob.result", "Sesión de Bob cerrada", data={
                "status": raw.get("status"),
                "cost": stats.get("session_costs"),
                "max_cost": stats.get("max_cost"),
                "duration_ms": stats.get("duration_ms"),
                "tool_calls": stats.get("tool_calls"),
            })

    def _report(self, raw: dict[str, Any]) -> None:
        """Informe que un subagente devuelve al orquestador: primera frase y tamaño, sin volcarlo."""
        agent = self._spawned.pop(str(raw.get("tool_id")))
        output = _TASK_TAGS.sub("", str(raw.get("output", ""))).strip()
        first = re.split(r"(?<=[.!?])\s|\n", output, maxsplit=1)[0] if output else ""
        self._emit(raw, "bob.subagent.report", f"{agent} entregó su informe al orquestador",
                   _clip(_strip_paths(first, self.workspace), MAX_REPORT),
                   {"agent": agent, "chars": len(output), "status": raw.get("status")}, actor=agent)

    def _tool(self, raw: dict[str, Any]) -> None:
        name = str(raw.get("tool_name", "herramienta"))
        params = raw.get("parameters") or {}
        path = relative_to_workspace(str(params.get("path", "")), self.workspace) if params.get("path") else None
        if name == "update_todo_list":
            items = parse_todos(_strip_paths(str(params.get("todos", "")), self.workspace))
            active = next((item["text"] for item in items if item["state"] == "active"), None)
            self._emit(raw, "bob.plan", "Bob actualizó su plan", active, {"items": items})
        elif name == "spawn_subagent":
            # El inicio en tiempo real llega del log de Bob (subagent_start); aquí solo se asocia el
            # tool_id para reconocer después el informe que el subagente devuelve al orquestador.
            self._spawned[str(raw.get("tool_id"))] = str(params.get("name", "subagente"))
        elif name == "use_skill":
            skill = str(params.get("skill_name", ""))
            self._emit(raw, "bob.skill", f"Activó la skill {skill}", data={"skill": skill})
        elif name == "read_file":
            self._emit(raw, "bob.tool", f"Leyó {path}", data={"tool": name, "target": path})
        elif name == "list_files":
            self._emit(raw, "bob.tool", f"Listó {path or '.'}", data={"tool": name, "target": path or "."})
        elif name == "glob":
            pattern = _clip(str(params.get("pattern", "")), 120)
            self._emit(raw, "bob.tool", f"Buscó archivos {pattern}", data={"tool": name, "target": pattern})
        elif name in ("search_files", "grep", "codebase_search"):
            query = _clip(str(params.get("regex") or params.get("query") or params.get("pattern") or ""), 120)
            self._emit(raw, "bob.tool", f"Buscó «{query}»" + (f" en {path}" if path else ""), data={"tool": name, "target": query})
        elif name in ("write_to_file", "apply_diff", "insert_content", "search_and_replace", "edit_file", "edit"):
            self._emit(raw, "bob.edit", f"Escribió {path}" if name == "write_to_file" else f"Editó {path}", data={"tool": name, "target": path})
        else:
            self._emit(raw, "bob.tool", f"Usó {name}", data={"tool": name, "target": path})
